src/OpenTelemetry/Logs/LogRecord.cs

```diff
@@ -213,22 +221,77 @@ public void ForEachScope<TState>(Action<LogRecordScope, TState> callback, TState
             return ref this.Data;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void ResetReferenceCount()
+        {
+            this.PoolReferenceCount = 1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void AddReference()
+        {
+            Interlocked.Increment(ref this.PoolReferenceCount);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal int RemoveReference()
+        {
+            return Interlocked.Decrement(ref this.PoolReferenceCount);
+        }
+
+        internal void Buffer()
+        {
+            this.BufferLogStateValues();
+            this.BufferLogScopes();
+        }
+
+        internal LogRecord Copy()
+        {
+            this.Buffer();
+
+            return new()
+            {
+                Data = this.Data,
+                State = this.State,
+                StateValues = this.StateValues == null ? null : new List<KeyValuePair<string, object?>>(this.StateValues),
+                BufferedScopes = this.BufferedScopes == null ? null : new List<object?>(this.BufferedScopes),
+            };
+        }
+
+        /// <summary>
+        /// Buffers the state values attached to the log into a list so that
+        /// they can be safely processed after the log message lifecycle has
+        /// ended.
+        /// </summary>
+        private void BufferLogStateValues()
+        {
+            var stateValues = this.StateValues;
+            if (stateValues == null || stateValues == this.AttributeStorage)
+            {
+                return;
+            }
+
+            var attributeStorage = this.AttributeStorage ??= new List<KeyValuePair<string, object?>>(stateValues.Count);
+            attributeStorage.AddRange(stateValues);
+            this.StateValues = attributeStorage;
+        }
+
         /// <summary>
         /// Buffers the scopes attached to the log into a list so that they can
         /// be safely processed after the log message lifecycle has ended.
         /// </summary>
-        internal void BufferLogScopes()
+        private void BufferLogScopes()
         {
-            if (this.ScopeProvider == null || this.bufferedScopes != null)
+            if (this.ScopeProvider == null)
             {
                 return;
             }
 
-            List<object?> scopes = new List<object?>();
+            List<object?> scopes = this.BufferedScopes ??= new List<object?>(16);
```

Comment:
Why 16?

Reply:
Okay. I learnt that it's the maximum number of scopes allowed. Maybe use: `LogRecordSharedPool.DefaultMaxNumberOfScopes`?

Reply:
Also, how is the list capacity restricted to 16? `AddScopeToBufferedList` is simply adding the scope to the list. It will just resize the list if it keeps adding more scopes to it, right?

Reply:
I think my naming might have confused things a bit. There isn't a maximum per `LogRecord` being processed. If a `LogRecord` has 1000 attributes and 100 scopes, they will all be added and handed to the exporter. What the "max" values on the pool control is what we'll keep around in the pool. So that `LogRecord` with 1000/100 will have its attributes/scopes `null`ed out when returned to the pool. Essentially the pool is constrained but individual `LogRecord`s are not.

Why 64/16? I just pulled those numbers out of the air. Tried to come up with something reasonable that would fit most use cases.

I did update to use the constant here. Let me know if given the above explanation you think other changes could help.

Reply:
I understood it now. I can't think of any other changes that could help other than maybe just renaming the variable to something like `DefaultInitialNumberOfScopes`.